File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""Defines a rectangle class."""


class Rectangle:
    """A rectangle class representation."""
    def __init__(self, width=0, height=0):
        """Initialize rectangle properties.

        Args:
            width (int). The width of the rectangle
            height (int). The height of the rectangle
        """
        self.__width = width
        self.__height = height

    @property
    def width(self):
        """Get the width of the current rectangle."""
        return self.__width

    @width.setter
    def width(self, width):
        """Set the width of the current rectangle
            depending on type and value
        """
        if not isinstance(width, int):
            raise TypeError('width must be an integer')
        elif width < 0:
            raise ValueError('width must be >= 0')
        self.__width = width

    @property
    def height(self):
        """Get the height of the current rectangle."""
        return self.__height

    @height.setter
    def height(self, height):
        """Set the height of the current rectangle
            depending on type and value
        """
        if not isinstance(height, int):
            raise TypeError('height must be an integer')
        elif height < 0:
            raise ValueError('height must be >= 0')
        self.__height = height

    def area(self):
        """Returns rectangle area."""
        return self.__width * self.__height

    def perimeter(self):
        """Returns rectangle perimeter."""
        if self.__width == 0 or self.__height == 0:
            return 0
        return (self.__width + self.__height) * 2

    def __str__(self):
        """Returns string representation of the rectangle."""
        if self.__width == 0 or self.__height == 0:
            return ('')
        rectangle = []
        for i in range(self.__height):
            for j in range(self.__width):
                rectangle.append('#')
            if i != self.__height - 1:
                rectangle.append("\n")
        return ("".join(rectangle))

    def __repr__(self):
        """Returns string representation of the rectangle."""
        rep = "Rectangle(" + str(self.__width) + "," + str(self.__height) + ")"
        return (rep)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print () {
    this.rect = [];
    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        this.rect.push('X');
      }
      if (i !== this.height - 1) { this.rect.push('\n'); }
    }
    console.log(this.rect.join(''));
  }

  rotate () {
    const temp = this.width;
    this.width = this.height;
    this.height = temp;
  }

  double () {
    this.width *= 2;
    this.height *= 2;
  }
}

module.exports = Rectangle;
